src/sync.cpp

```diff
@@ -1950,5 +1959,575 @@ bool Sync::movetolocaldebris(LocalPath& localpath)
+            mClient.app->sync_removed(tag);
+        }
+    }
+}
+
+void Syncs::disableSyncs(SyncError syncError, bool newEnabledFlag)
+{
+    bool anySyncDisabled = false;
+    disableSelectedSyncs([&](SyncConfig&, Sync* s){
+
+        if (s)
+        {
+            anySyncDisabled = true;
+            return true;
+        }
+        return false;
+    }, syncError, newEnabledFlag);
+
+    if (anySyncDisabled)
+    {
+        LOG_info << "Disabled syncs. error = " << syncError;
+        mClient.app->syncs_disabled(syncError);
+    }
+}
+
+
+void Syncs::disableSelectedSyncs(std::function<bool(SyncConfig&, Sync*)> selector, SyncError syncError, bool newEnabledFlag)
+{
+    for (auto i = mSyncVec.size(); i--; )
+    {
+        if (selector(mSyncVec[i]->mConfig, mSyncVec[i]->mSync.get()))
+        {
+            if (auto sync = mSyncVec[i]->mSync.get())
+            {
+                sync->changestate(SYNC_DISABLED, syncError, newEnabledFlag, true); //This will cause the later deletion of Sync (not MegaSyncPrivate) object
+                mClient.syncactivity = true;
+            }
+            else
+            {
+                mSyncVec[i]->mConfig.setError(syncError);
+                mSyncVec[i]->mConfig.setEnabled(newEnabledFlag);
+                mSyncVec[i]->changedConfigState(true);
+            }
+        }
+    }
+}
+
+
+void Syncs::removeSelectedSyncs(std::function<bool(SyncConfig&, Sync*)> selector)
+{
+    for (auto i = mSyncVec.size(); i--; )
+    {
+        if (selector(mSyncVec[i]->mConfig, mSyncVec[i]->mSync.get()))
+        {
+            removeSyncByIndex(i);
+        }
+    }
+}
+
+void Syncs::removeSyncByIndex(size_t index)
+{
+    if (index < mSyncVec.size())
+    {
+        if (auto& syncPtr = mSyncVec[index]->mSync)
+        {
+            syncPtr->changestate(SYNC_CANCELED, UNKNOWN_ERROR, false, false);
+
+            if (syncPtr->statecachetable)
+            {
+                syncPtr->statecachetable->remove();
+                delete syncPtr->statecachetable;
+                syncPtr->statecachetable = NULL;
+            }
+            syncPtr.reset(); // deletes sync
+        }
+
+        // call back before actual removal (intermediate layer may need to make a temp copy to call client app)
+        auto tag = mSyncVec[index]->mConfig.getTag();
+        mClient.app->sync_removed(tag);
+
+        mSyncConfigDb->removeByTag(tag);
+        mClient.syncactivity = true;
+        mSyncVec.erase(mSyncVec.begin() + index);
+
+        isEmpty = mSyncVec.empty();
+    }
+}
+
+
+error Syncs::enableSyncByTag(int tag, bool resetFingerprint, UnifiedSync*& syncPtrRef)
+{
+    for (auto& s : mSyncVec)
+    {
+        if (s->mConfig.getTag() == tag)
+        {
+            syncPtrRef = s.get();
+            if (!s->mSync)
+            {
+                return s->enableSync(resetFingerprint, true);
+            }
+            s->mConfig.setError(ACTIVE_SYNC_BELOW_PATH);
+            s->mConfig.setEnabled(false);
```

Comment:
A good point - actually I'm not sure where these two lines came from, I think I must have copied them from somewhere erroneously.   The case here is that the app asked to enable a sync, but it's already running.  We shouldn't interfere with the already running sync or it's config, we can just return EEXIST.   updated, thanks

Reply:
Actually the tests were checking for ACTIVE_SYNC_BELOW_PATH in this case.  I've modified that to NO_ERROR.   I think originally that value was not stored in the config, but passed around in function parameters.  However, it's not clear from the megaapi.h documentation as to how the error should be interpreted - I would have thought any error passed back represented the state of the sync.    Another possibility would be to not consider this an error even for the return code (ie, not return API_EEXIST) - after all, the sync is enabled already so the calling app should be satisfied.

Reply:
These cases are always tricky. I tend to prefer the error (`API_EEXIST`), since it provides extra information to the app about what's actually going on, which is positive for a nice UX. Imagine we silently return: all good. Then, the app could notify: "Your sync has been enabled" and do anything else that should ONLY happen when a sync has been enabled.
Instead, the app should say: "Your sync is already enabled" and nothing else. So good to go as it is now, by returning `API_EEXIST`, IMO.